src/wp-includes/class-wp-theme-json.php

```diff
@@ -0,0 +1,383 @@
+<?php
+/**
+ * Process of structures that adhere to the theme.json schema.
+ *
+ * @package WordPress
+ */
+
+/**
+ * Class that encapsulates the processing of
+ * structures that adhere to the theme.json spec.
+ */
+class WP_Theme_JSON {
+
+	/**
+	 * Container of data in theme.json format.
+	 *
+	 * @var array
+	 */
+	private $theme_json = null;
+
+	const ALLOWED_TOP_LEVEL_KEYS = array(
+		'version',
+		'settings',
+	);
+
+	const ALLOWED_SETTINGS = array(
+		'color'      => array(
+			'custom'         => null,
+			'customGradient' => null,
+			'duotone'        => null,
+			'gradients'      => null,
+			'link'           => null,
+			'palette'        => null,
+		),
+		'custom'     => null,
+		'layout'     => null,
+		'spacing'    => array(
+			'customMargin'  => null,
+			'customPadding' => null,
+			'units'         => null,
+		),
+		'typography' => array(
+			'customFontSize'        => null,
+			'customLineHeight'      => null,
+			'dropCap'               => null,
+			'fontSizes'             => null,
+		),
+	);
+
+	const LATEST_SCHEMA = 1;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param array $theme_json A structure that follows the theme.json schema.
+	 */
+	public function __construct( $theme_json = array() ) {
+		if ( ! isset( $theme_json['version'] ) || self::LATEST_SCHEMA !== $theme_json['version'] ) {
+			$this->theme_json = array();
+			return;
+		}
+
+		$valid_block_names = array_keys( WP_Block_Type_Registry::get_instance()->get_all_registered() );
+		$this->theme_json  = self::sanitize( $theme_json, $valid_block_names );
+	}
+
+	/**
+	 * Returns the allowed schema for a theme.json structure.
+	 *
+	 * @return array
+	 */
+	private static function get_allowed_schema() {
+		return apply_filters(
+			'theme_json_allowed_schema',
```

Comment:
Initially, this was `theme_json_allowed_settings`. However, it makes more sense to have a single filter for the whole schema (settings, styles, etc).

An alternative implementation of the filter I've considered would be this:

```php
apply_filters(
  'theme_json_allowed_schema',
  array(
    'version'  => null,
    'settings' => array(
      'color' => array(),
      'typography' => array(),
      'blocks' => array(
        'color' => array(),
        'typography' => array(),
      )
    }
  )
);
```

I like more the current approach for a few reasons:

- What core does is explicit: filters out the top-level keys but only does some inner filtering for specific keys (such as `settings`). By being explicit, any future addition of more data (`topLevel`, `settings`, `styles`, etc) doesn't require the consumers to change anything. They can do `array_merge( $allowed_schema, array( 'settings' => array() ) );` and only the settings will be overwritten, the other keys will stay the same
- It doesn't require the consumer to know the shape of the `theme.json`. For example, the consumer only needs to provide the allowed settings and core takes care of applying that to every level (top or per block). This would help us to migrate the shape in the future, should we need it, as it's not exposed to consumers.

Reply:
Removed it as per https://github.com/WordPress/wordpress-develop/pull/1262#discussion_r636834064